### DISCUSSION TITLE
Angular - inject() must be called from an injection context

### DISCUSSION THREAD

**Discussion by A1Jan** [storybookjs/storybook#26612] [Help] [answered]
### Summary

Hello,

we are using Storybook with a custom preview.ts file to provide the global context:
```
const meta: Meta = {
    decorators: [
        applicationConfig({
            providers: [
                importProvidersFrom(RouterTestingModule, ...),
                ...
            ]
        })
    ]
};
```
A story looks like this:
```
const meta: Meta<TestInputComponent> = {
    component: TestInputComponent,
    decorators: [
        moduleMetadata({
            imports: [...some additional modules],
        }),
    ],
};
export default meta;

type Story = StoryObj<TestInputComponent>;

export const Default: Story = {
    render: (args) => ({
        template: `<test-input></test-input>`,
    })
};
```
In general this is working fine, but when we add one special plugin (import and use a directive) which is this one:
https://github.com/taiga-family/maskito/blob/main/projects/angular/src/lib/maskito.directive.ts

The story will not load and has an overlay:
```
Error: NG0203: inject() must be called from an injection context such as a constructor, a factory function, a field initializer, or a function used with `runInInjectionContext`. Find more at https://angular.io/errors/NG0203
  at injectInjectorOnly (/vendors-node_modules_pnpm_css-loader_6_8_1_webpack_5_89_0_node_modules_css-loader_dist_runtim-d772f2.iframe.bundle.js:11991:11)
  at ɵɵinject (/vendors-node_modules_pnpm_css-loader_6_8_1_webpack_5_89_0_node_modules_css-loader_dist_runtim-d772f2.iframe.bundle.js:12001:59)
  at inject (/vendors-node_modules_pnpm_css-loader_6_8_1_webpack_5_89_0_node_modules_css-loader_dist_runtim-d772f2.iframe.bundle.js:12084:10)
  at MaskitoDirective (/vendors-node_modules_pnpm_css-loader_6_8_1_webpack_5_89_0_node_modules_css-loader_dist_runtim-d772f2.iframe.bundle.js:52598:76)
  at MaskitoDirective_Factory (/vendors-node_modules_pnpm_css-loader_6_8_1_webpack_5_89_0_node_modules_css-loader_dist_runtim-d772f2.iframe.bundle.js:52641:10)
  at getNodeInjectable (/vendors-node_modules_pnpm_angular_common_17_0_5__angular_core_17_0_5_rxjs_7_8_1_node_modules_-104325.iframe.bundle.js:56071:38)
  at instantiateAllDirectives (/vendors-node_modules_pnpm_angular_common_17_0_5__angular_core_17_0_5_rxjs_7_8_1_node_modules_-104325.iframe.bundle.js:64230:40)
  at createDirectivesInstances (/vendors-node_modules_pnpm_angular_common_17_0_5__angular_core_17_0_5_rxjs_7_8_1_node_modules_-104325.iframe.bundle.js:63666:27)
  at ɵɵelementStart (/vendors-node_modules_pnpm_angular_common_17_0_5__angular_core_17_0_5_rxjs_7_8_1_node_modules_-104325.iframe.bundle.js:73268:30)
```

The storys will not load anymore. We are also using inject() in some of our components, but only this reference makes trouble. Is it because of the inject() call in the variables? Maybe someone has a useful hint for me what is going wrong here.

Thanks in advance.

**Comment by Marklb** [comment]
There shouldn't be any problem. Those `inject` calls on the instance variables would be in the injection context, unless you are using some non-standard way of initializing the directive.

The `preview.ts` should not affect it, but I can't say for sure since some of it is omitted.

Here is a Stackblitz that implements a story for the first example in their docs: https://stackblitz.com/edit/github-vojczh-vfrsqd

Could you provide more information? Ideally a reproduction, but at the least your Angular and Storybook versions?

Is your component standalone? Do you import `MaskitoDirective` or `MaskitoModule`? Where do you import `MaskitoDirective`/`MaskitoModule`?

**Reply by A1Jan** [reply]
Dependencies:
Angular 17.0.5
Storybook 8.0.2

We are using the ion-input component form Ionic https://ionicframework.com/docs/api/input with "Input masking", which uses the maskito plugin. As soon as I remove the maskito reference, the error is gone. My "TestComponent" from above is a wrapper component (standalone) which uses the "<ion-input [maskito...". The TestComponent has the import to the standalone directive of maskito. We used the MaskitoModule before with the same error.

Thanks for the stackblitz link, I will have a look if I can reproduce it there somehow.

**Reply by A1Jan** [reply]
Update: I tried to reproduce it with the example but everything worked, so I moved to my project and tried to some more research there.

The reason seems not to be the imports or module, but how or where my angular component is referenced. Our structure in this project is the following: The project started to be used as a shared library which provides components for a design system. So the setup is on one side a default angular project to setup storybook and additionally an angular library project which provides components like a shared input component, which can be published to a npm repository. The components directory is 'projects/components/src/lib/...' and the stories are placed in 'stories/...'. When I now copy the input component to my stories folder, everything works without an error, but when I reference it with `import { A1InputComponent } from 'projects/components/src/lib/input/input.component';` I see the error message.
![grafik](https://github.com/storybookjs/storybook/assets/161336748/a78910a2-7b63-4178-85a8-199cbefb985f)
And as soon as I remove the "[maskito]" attribute, the error is gone. Do you have any idea why this behavior is occuring?

I forked your stackblitz and added an angular library, but of course the problem is not occuring there... Maybe I just have to build my code up step by step from zero.

**Reply by A1Jan** [reply, marked as answer]
The solution seems to be the paths property inside the main tsconfig.ts
```
{
    ...
    "compilerOptions": {
        "paths": {
            "@my/libraryname": ["./dist/angular-library"]
        }
        ...
    }
    ...
}
```
Now I can reference my component with '@my/libraryname' and the error is gone.

**Reply by Marklb** [reply]
I use the paths option to set a path for my libraries, but I wouldn't think you would want to point to the `dist` directory. You would have to precompile your components, which is going to also strip out a lot of the info for auto generating controls. The generation may still work though, because of how Compodoc is current used.

I have also run into problems with paths from the root, like `from 'projects/components/src/lib...'`. Instead, I try to stick to relative paths and they usually work. My library project is not configured exactly normal and needs to be reorganized, but I have something like the following in my root tsconfig `{..., "compilerOptions": { "paths": "@my/libraryname/*": [ "projects/libraryname/*/public-api" ] }}`. I use the `*`, because I have it configured with multiple exports. If you go with that, it may be like the following for you `{..., "compilerOptions": { "paths": "@my/libraryname": [ "projects/libraryname/src/lib/public-api" ] }}`.

**Reply by A1Jan** [reply]
Your absolutely correct. I have the same case with multiple libraries and also changed it like this afterwards to prevent to compile after each change. Thanks for your hints and your help! My mentioned code above is the default configuration when adding an angular library.
